onnxruntime/core/providers/cpu/math/inverse.cc

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/platform/threadpool.h"
+#include "core/util/math_cpuonly.h"
+#include "Eigen/src/Core/Map.h"
+#include "Eigen/LU"
+#include <functional>
+
+namespace onnxruntime {
+class Inverse final : public OpKernel {
+ public:
+  explicit Inverse(const OpKernelInfo& info) : OpKernel(info) {}
+  Status Compute(OpKernelContext* ctx) const override;
+
+ private:
+  template <typename T>
+  struct ComputeImpl;
+};
+
+ONNX_CPU_OPERATOR_KERNEL(
+    Inverse,
+    12,
+    KernelDefBuilder()
+        .TypeConstraint("T", BuildKernelDefConstraints<float, double, MLFloat16>()),
+    Inverse);
+
+template <typename T>
+using MatrixT = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+
+template <typename T>
+struct Inverse::ComputeImpl {
+  void operator()(const Tensor* input, Tensor* output,
+                  int64_t batch_num, int64_t rows, int64_t cols) const {
+    auto batch_offset = batch_num * rows * cols;
+    const auto* input_data = input->Data<T>() + batch_offset;
+    auto* output_data = output->MutableData<T>() + batch_offset;
+
+    Eigen::Map<const MatrixT<T>> input_matrix(input_data, rows, cols);
+    Eigen::Map<MatrixT<T>> output_matrix(output_data, rows, cols);
+    output_matrix = input_matrix.inverse();
+  }
+};
+
+template <>
+struct Inverse::ComputeImpl<MLFloat16> {
+  void operator()(const Tensor* input, Tensor* output,
+                  int64_t batch_num, int64_t rows, int64_t cols) const {
+    auto batch_offset = batch_num * rows * cols;
+    // Direct cast to half as it just as MLFloat16 containes only uint16_t
+    const auto* input_data = reinterpret_cast<const Eigen::half*>(input->Data<MLFloat16>() + batch_offset);
+    auto* output_data = reinterpret_cast<Eigen::half*>(output->MutableData<MLFloat16>() + batch_offset);
+
+    Eigen::Map<const MatrixT<Eigen::half>> input_matrix(input_data, rows, cols);
+    Eigen::Map<MatrixT<Eigen::half>> output_matrix(output_data, rows, cols);
+    output_matrix = input_matrix.inverse();
```

Comment:
Does Eigen have a limit on the size of the matrix's rows (and cols) by any chance ?

Reply:
This indicates ([https://stackoverflow.com/questions/17430644/what-is-the-maximum-size-of-matrix-in-eigen](https://stackoverflow.com/questions/17430644/what-is-the-maximum-size-of-matrix-in-eigen)) that it is bound by memory available on the system.

---
In reply to: [407768947](https://github.com/microsoft/onnxruntime/pull/3485#discussion_r407768947) [](ancestors = 407768947)